src/test/java/jenkins/plugins/logstash/LogstashWriterTest.java

```diff
@@ -0,0 +1,299 @@
+    // Verify results
+    // No error output
+    assertEquals("Results don't match", "", errorBuffer.toString());
+    verify(mockBuild).getLog(3);
+
+    verify(mockDao).buildPayload(Matchers.eq(mockBuildData), Matchers.eq("http://my-jenkins-url"), Matchers.anyListOf(String.class));
+    verify(mockDao).push("{\"data\":{},\"message\":[\"test\"],\"source\":\"jenkins\",\"source_host\":\"http://my-jenkins-url\",\"@version\":1}");
+  }
+
+  @Test
+  public void writeSuccessConnectionBroken() throws Exception {
+    Mockito.doNothing().doThrow(new IOException("BOOM!")).doNothing().when(mockDao).push(anyString());
+    LogstashWriter los = createLogstashWriter(mockBuild, errorBuffer, "http://my-jenkins-url", mockDao, mockBuildData);
+
+
+    String msg = "test";
+    String exMessage = "[logstash-plugin]: Failed to send log data to REDIS:localhost:8080.\n" +
+      "[logstash-plugin]: No Further logs will be sent to localhost:8080.\n" +
+      "java.io.IOException: BOOM!\n";
+
+    errorBuffer.reset();
+
+    // Unit under test
+    los.write(msg);
+
+    // Verify results
+    assertEquals("Results don't match", "", errorBuffer.toString());
+
+    // Break the dao connnection
+    errorBuffer.reset();
+
+    // Unit under test
+    los.write(msg);
+
+    // Verify results
+    assertTrue("Results don't match", errorBuffer.toString().startsWith(exMessage));
+    assertTrue("Connection not broken", los.isConnectionBroken());
+
+    // Verify logs still write but on further calls are made to dao
+    errorBuffer.reset();
+    // Unit under test
+    los.write(msg);
+
+    // Verify results
+    assertEquals("Results don't match", "", errorBuffer.toString());
+
+    //Verify calls were made to the dao logging twice, not three times.
+    verify(mockDao, times(2)).buildPayload(Matchers.eq(mockBuildData), Matchers.eq("http://my-jenkins-url"), Matchers.anyListOf(String.class));
+    verify(mockDao, times(2)).push("{\"data\":{},\"message\":[\"test\"],\"source\":\"jenkins\",\"source_host\":\"http://my-jenkins-url\",\"@version\":1}");
+    verify(mockDao).getIndexerType();
+    verify(mockDao, times(2)).getDescription();
+  }
+
+  @Test
+  public void writeBuildLogGetLogError() throws Exception {
+    // Initialize mocks
+    when(mockBuild.getLog(3)).thenThrow(new IOException("Unable to read log file"));
+
+    LogstashWriter writer = createLogstashWriter(mockBuild, errorBuffer, "http://my-jenkins-url", mockDao, mockBuildData);
+    assertEquals("Errors were written", "", errorBuffer.toString());
+
+    // Unit under test
+    writer.writeBuildLog(3);
+
+    // Verify results
+    verify(mockBuild).getLog(3);
+
+    String errorMsg = "[logstash-plugin]: Unable to serialize log data.\n" +
+      "java.io.IOException: Unable to read log file\n";
+    assertThat("Wrong error message", errorBuffer.toString(), containsString(errorMsg));
+    verify(mockDao).buildPayload(Matchers.eq(mockBuildData), Matchers.eq("http://my-jenkins-url"), anyListOf(String.class));
```

Comment:
In this case, the errorMsg should also be pushed to logstash (the connection is still active, we should report that we weren't able to get the real log).  But I can't figure out how to do it.   `ArgumentCaptor` for the `logLines` list?   Suggestions?  

It can be a follow on commit, but I'd like to know how to do it.